### PR TITLE
[ML][Pipelines] redo feat: flow in pipeline

### PR DESCRIPTION
**Pull request by elliotzh**
# Description

This reverts PR https://github.com/Azure/azure-sdk-for-python/pull/31911 and apply a fix to remove `--extra-index-url` in requirements.txt file to pass supply chain analysis.

Follow-up issue to prevent this in the future is filed at https://github.com/Azure/azure-sdk-for-python/issues/31910

Manually triggered build containing supply chain analysis: [link](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3049789&view=results)

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by wangchao1230 — APPROVED** [review]

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml [hunk 1]
```diff
@@ -26,8 +26,8 @@ repos:
     args: ['--settings-file', 'sdk/ml/azure-ai-ml/pyproject.toml']
 - repo: local
   hooks:
-    - id: pylint
-      name: pylint
+    - id: pylint-dependencies-check
+      name: pylint-dependencies-check
```
Comment (kdestin) [inline sdk/ml/azure-ai-ml/.pre-commit-config.yaml:30]: What's the rationale behind splitting this into two steps?

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml [hunk 2]
```diff
@@ -69,9 +69,9 @@ repos:
...
-        os.execl(sys.executable, sys.executable, "-m", "pylint", *sys.argv[1:])
-      # cspell:disable-next-line
-      - "--rcfile=pylintrc"
-      - "--output-format=parseable"
+    - id: pylint
+      name: pylint
+      entry: pylint
+      language: system
+      args: [--rcfile=pylintrc, --output-format=parseable]
+      files: '^sdk/ml/azure-ai-ml/azure/ai/ml/.*\.py$'
```
Comment (kdestin) [inline sdk/ml/azure-ai-ml/.pre-commit-config.yaml:77]:
```suggestion
      types: [python]
```

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml [hunk 3]
```diff
@@ -69,9 +69,9 @@ repos:
...
-        # Run pylint
-        os.execl(sys.executable, sys.executable, "-m", "pylint", *sys.argv[1:])
-      # cspell:disable-next-line
-      - "--rcfile=pylintrc"
-      - "--output-format=parseable"
+    - id: pylint
+      name: pylint
+      entry: pylint
+      language: system
+      args: [--rcfile=pylintrc, --output-format=parseable]
```
Comment (kdestin) [inline sdk/ml/azure-ai-ml/.pre-commit-config.yaml:76]:
```suggestion
      entry: python
      language: system
      args: [-m, pylint, --rcfile=pylintrc, --output-format=parseable]
```

**Review by kdestin — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml (hunk 1 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/.pre-commit-config.yaml:30]: the `pylint` in combined check has some error in applying a correct rcfile in some cases

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml (hunk 2 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/.pre-commit-config.yaml]: applied

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml (hunk 3 above)
Comment (elliotzh) [reply sdk/ml/azure-ai-ml/.pre-commit-config.yaml]: applied

File: sdk/ml/azure-ai-ml/.pre-commit-config.yaml (hunk 1 above)
Comment (kdestin) [reply sdk/ml/azure-ai-ml/.pre-commit-config.yaml:30]: I'm not exactly sure how splitting this into two steps would resolve finding the correct rcfile, but I'm guessing I'm missing something.

---

I'm not interested in blocking your PR over this, so feel free to merge if you need to.

Comment (elliotzh) [reply sdk/ml/azure-ai-ml/.pre-commit-config.yaml:30]: Thanks, at least it won't fail for now. Will involve you next time I'm blocked by this

**Review by kdestin — APPROVED** [review]

**Pull request merged**